port/unix/omrsysinfo.c

```diff
@@ -6775,12 +6890,45 @@ omrsysinfo_cgroup_subsystem_iterator_next(struct OMRPortLibrary *portLibrary, st
 		 */
 		size_t len = strlen(metricElement->value);
 		metricElement->units = currentElement->metricUnit;
-		if ((len > 0) && (metricElement->value[len-1] == '\n')) {
+		if ((len > 1) && (metricElement->value[len-1] == '\n')) {
 			metricElement->value[len-1] = '\0';
 		}
 		if (currentElement->isValueToBeChecked) {
 			int64_t result = 0;
-			sscanf(metricElement->value, "%" PRId64, &result);
+			if ('\n' == metricElement->value[0]) {
+				/* Empty file, see cpuset.cpus and cpuset.mems in omrCgroupCpusetMetricMapV2. */
+				result = -1;
+			} else if (OMR_ARE_ANY_BITS_SET(PPG_sysinfoControlFlags, OMRPORT_SYSINFO_CGROUP_V1_AVAILABLE)) {
+				rc = sscanf(metricElement->value, "%" PRId64, &result);
+				if (1 != rc) {
+					rc = portLibrary->error_set_last_error_with_message_format(
+							portLibrary,
+							OMRPORT_ERROR_SYSINFO_CGROUP_SUBSYSTEM_FILE_INVALID_VALUE,
+							"invalid value %s in file %s",
+							metricElement->value,
+							subsystemMetricMapElement->metricFileName);
+				} else {
+					rc = 0;
+				}
+			} else if (OMR_ARE_ANY_BITS_SET(PPG_sysinfoControlFlags, OMRPORT_SYSINFO_CGROUP_V2_AVAILABLE)) {
+				uint64_t uresult = 0;
+				rc = scanCgroupIntOrMax(portLibrary, metricElement->value, &uresult);
+				if (0 != rc) {
+					rc = portLibrary->error_set_last_error_with_message_format(
+							portLibrary,
+							OMRPORT_ERROR_SYSINFO_CGROUP_SUBSYSTEM_FILE_INVALID_VALUE,
+							"invalid value %s in file %s",
+							metricElement->value,
+							subsystemMetricMapElement->metricFileName);
+				} else if (UINT64_MAX == uresult) {
+					result = -1;
+				} else {
+					result = (int64_t)uresult;
```

Comment:
If overflowed, `result` will be negative and the value will be "Not set". Should I make this explicit by `(UINT64_MAX == uresult) || (uresult > INT64_MAX)`?

Reply:
Will there be an overflow for a valid result? Can valid results be greater than INT64_MAX?

Reply:
```
root@purpurin1 ~/d/openj9-openjdk-jdk8 (openj9)# echo 1000000000000000000 > /sys/fs/cgroup/user.slice/user-0.slice/session-876.scope/memory.max
root@purpurin1 ~/d/openj9-openjdk-jdk8 (openj9)# cat /sys/fs/cgroup/user.slice/user-0.slice/session-876.scope/memory.max
1000000000000000000
```
But in docker:
```
root@purpurin1 ~/d/openj9-openjdk-jdk8 (openj9)# docker run -v /root/docker_hostdir:/root/hostdir -it -m 1000000000000000GB --cpu-period=1000 --cpu-quota=10000 --cpuset-cpus=1,2,4-5  --rm openj9
docker: Error response from daemon: Minimum memory limit allowed is 6MB.
See 'docker run --help'.
```
Seems like docker mem limit overflowed

Reply:
> If overflowed, result will be negative and the value will be "Not set". Should I make this explicit by (UINT64_MAX == uresult) || (uresult > INT64_MAX)?

For valid values (greater than INT64_MAX), we should not translate it to "Not Set".

Reply:
Maybe
```
} else if (UINT64_MAX == uresult) {
	result = -1;
} else if (uresult > INT64_MAX) {
        result = 0;
}else {
	result = (int64_t)uresult;
}
```
This will skip the following check and leave the metricElement->value string as is

Reply:
Then, let's do it for all values less than `UINT64_MAX`: (uresult < UINT64_MAX)`. With this change, I do not think the `else` statement will be invoked.

Reply:
Line 6932: `(result > (MAX_DEFAULT_VALUE_CHECK)`
```
#define MAX_DEFAULT_VALUE_CHECK (LLONG_MAX - (1024 * 1024 * 1024)) /* subtracting the MAX page size (1GB) from LLONG_MAX to check against a value */
```
`UINT64_MAX` is less than `MAX_DEFAULT_VALUE_CHECK`. Setting `result = 0` for `(uresult < UINT64_MAX)` won't be in conflict with the check on Line 6932.

Reply:
> UINT64_MAX is less than MAX_DEFAULT_VALUE_CHECK

Do you mean greater?
Also can't we just do
```
} else if (UINT64_MAX == uresult) {
	result = -1;
} else {
        result = 0;
}

Reply:
The `(result > (MAX_DEFAULT_VALUE_CHECK))` check sets "Not set" for all values larger than the max of 64bit signed int minus 2^30. Setting result = 0 for uresult < UINT64_MAX will conflict with this. e.g. for uresult = LLONG_MAX

Reply:
> So the original code makes sense since all values greater than max signed 64 bits will go negative and will be set to "Not set".

This will only be true if `MAX_DEFAULT_VALUE_CHECK` applies to cgroup v2 and the values returned by `scanCgroupIntOrMax`. Can you verify this detail?

Reply:
I don't think MAX_DEFAULT_VALUE_CHECK is related to cgroups, it is probably just a sanity check that any limit above it is effectively unlimited.

Reply:
>  So the original code makes sense since all values greater than max signed 64 bits will go negative and will be set to "Not set".

The original code is implicit, which may be incorrectly interpreted. We should make an explicit change and add a comment to justify our changes. e.g.
```
} else if (uresult > MAX_DEFAULT_VALUE_CHECK) { // This should also cover (UINT64_MAX == uresult).
        /* It is assumed that the metric is effectively unlimited, and can be considered "Not Set". */
	result = -1;
} else {
        result = 0;
}
```

Reply:
Added changes